### PR TITLE
Dynamic primary zones: API-provisioned, template-constrained primaries

### PR DESCRIPTION
**Pull request by johanix**
Implements the agreed design in `docs/2026-07-13-dynamic-primary-zones.md`: API-provisioned, **template-constrained primary zones** — create/serve new primaries at runtime via `zone add --type primary --template <name>`, constrained to operator-blessed templates, persisted so a restart rebuilds them. One branch end-to-end, commits sequenced per the doc's §10.4 PR1→PR4 stages.

## What's in here

**Gates (§3)**
- `dynamiczones.dynamic.allowed` changes **bool → list of zone types** (`allowed: [secondary, primary]`), in its own `DynamicApiZoneConf` (catalog blocks keep the bool). A legacy bool is a **hard config error naming the new syntax** (decode hook wired into both the daemon loader and `config check`); unknown values are config errors; absent/empty = deny all. **This is the one breaking config change** — deployed configs using `dynamic: allowed: true/false` need the list syntax.
- Templates gain a `dynamiczones: true` blessing flag (never copied to zones). API adds require it; already-persisted zones load without it (blessing gates new adds, mirroring how `allowed:` treats persisted zones).

**Shared activation (§10.2-E)** — the update-policy validation/activation block was factored out of `ParseZones` into `activateUpdatePolicy`, behavior-preserving (error strings pinned by tests), and is used by ParseZones + the add path + the boot path, so they cannot drift.

**Primary add path (§4)** — `prepareDynamicPrimary` runs the same validation sequence ParseZones applies (alias-conflict, peer-refs, downstream-auth + ACLs, publish-cadence, DNSSEC-policy resolution, options, policy activation), bootstraps the apex (SOA serial 1, timers `3600 600 604800 300`, NS + glue from `dnsengine.addresses`; wildcard listeners skipped with WARN; existing files used as-is), registers a fully-configured first-load stub so the RefreshEngine runs the standard first-bind sequence — the bootstrapped apex lands through the post-snapshot publish path (§10.3 risk 2). Inline TSIG is reinterpreted for the primary role: keyless `downstreams` entries are rewired to the staged key (BLOCKED never touched).

**Restart (§5)** — the persisted entry carries `template:` + `type: primary`; `LoadDynamicZoneFiles` re-expands via the same shared helper (persisted fields win by gap-fill; update policy deliberately NOT persisted — re-derives from the template). Missing/broken template ⇒ visible ERROR-state zone, not a silent skip.

**Lifecycle** — `modify` refused for primaries (v1, per §5); `delete` also removes the template-path zone file; RFC 2136 ZONE-UPDATEs on an API-managed primary persist the zone file immediately (mirror of the CHILD-UPDATE `direct`-backend persist) so updated content survives a restart.

## Design-doc corrections + resolved forks

The doc's §2 "already exists" claims were re-verified against main @ 7885f23; two had drifted and are corrected in place:
- **Restart works via `LoadDynamicZoneFiles`, not `include:`+ParseZones** — the include merge *overrides* list-valued keys (an included dynamic file's `zones:` clobbers other files' zones) and the ParseZones route drops the ApiManaged marker.
- **Dirty-primary content was never auto-persisted** (write-back only ran on refresh, which dirty primaries refuse) — closed by the update-time write-back.

Every fork resolved during the (owner-AFK) implementation is recorded with rationale in the doc's new **§11** — highlights: v1 refuses `allow-child-updates` / catalog / multi-provider options in blessed templates; unusable dnssecpolicy refuses the add; caller-supplied options/primaries refused for primaries (the template IS the envelope); `zoneDataToZoneConf` honors `zd.Zonefile`; template reads under `confMu.RLock`.

## Found & fixed in live testing

A pre-existing latent bug affecting **all** dynamic zones: on a server with zero static zones, the default query handler was never registered, so every dynamically-added zone's queries got REFUSED (masked on testbeds that always carry static zones). The handler now also registers when dynamic zones are possible. Relatedly, `CheckDynamicConfigFileIncluded` warned in the wrong direction (pre-`LoadDynamicZoneFiles` leftover) — it now warns when the dynamic file IS included, and the sample config no longer lists it under `include:`.

## Test evidence

- Unit: gate matrix (allowed-list × blessing × type), legacy-bool config error, activation-helper behavior pinning, bootstrap edge cases (wildcard/dedupe/zero-address), TSIG downstreams rewiring (BLOCKED untouched), boot re-expansion incl. ERROR-state registration, modify-refusal, delete file removal, existing-file preservation.
- `go test -race ./...` green in all five modules (v2, v2/cli, v2/core, v2/cache, v2/edns0); gofmt clean on touched files; all six cmdv2 binaries build via make.
- **Live smoke** (fresh tdns-auth, empty `zones:`): add → SOA/NS/glue served from the synthesized apex → unsigned UPDATE refused → restart → template re-expanded, zone Ready and still update-gated → delete → zone, zone file, and dynamic-config entry all gone.

Not live-exercised: signed-UPDATE round-trip (needs SIG(0) truststore setup) and the inline-TSIG transfer gating — both unit-covered; suggest exercising via tdns-debug per §6 as follow-up.

## Summary by CodeRabbit

* **New Features**
  * Added API/CLI support for creating dynamic primary and secondary zones, with template selection for primary zones.
  * Introduced per-zone-type configuration controls for which dynamic zone types can be created via the API.
  * Dynamic primary provisioning now includes template-aware bootstrapping and refresh handling, including delegation sync when enabled.
* **Bug Fixes**
  * Improved validation and clearer errors for invalid dynamic-zone configuration (including legacy syntax handling).
  * Ensured API-managed primary zones persist successful primary updates promptly.

### CODE REVIEW

File: v2/structs.go [hunk 1]
```diff
@@ -700,14 +711,22 @@ type ZoneRefresher struct {
...
+	// not re-parse; zero means "leave the zone's default".
+	PublishCadence time.Duration
+	Options        map[ZoneOption]bool
+	Edns0Options   *edns0.MsgOptions
+	UpdatePolicy   UpdatePolicy
+	DnssecPolicy   string
+	MultiSigner    string
+	Force          bool // force refresh, ignoring SOA serial
+	Wait           bool // wait for refresh to complete before responding
+	Response       chan RefresherResponse
```
[the remark this answers is not in the document]
Comment (johanix) [reply v2/structs.go:750]: Declined, with reasoning: the existing-zone reload branch never touches `zd.Template`, so a `zone reload` cannot drop it — the field survives untouched and the next persist serializes it correctly. No current sender targets an *existing* zone with a config-bearing refresher carrying `Template`/`PublishCadence` (`LoadDynamicZoneFiles` skips zones already in the map, and static-zone reload goes through `ParseZones`, which never handles dynamic primaries). The suggested unconditional copy would itself introduce the clearing bug this finding warns about: CLI `zone reload` refreshers carry an empty `Template`, which would erase the field before the next persistence. A guarded only-if-provided copy would be dead code today; if a re-expansion-on-reload path for dynamic primaries appears later, the guarded merge belongs in that change. All other findings in this round were addressed in 7472548.
